File: frontend/my-app/src/api/habits.ts
import { apiRequest } from "./client";
import type { CreateHabitRequest, Habit, UpdateHabitRequest } from "./types";

export async function getHabits() {
  try {
    return await apiRequest<Habit[]>("/api/habits/");
  } catch (error) {
    if (error instanceof Error && error.message.includes("404")) {
      return [];
    }

    throw error;
  }
}

export function createHabit(payload: CreateHabitRequest) {
  return apiRequest<Habit>("/api/habits/create", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export function updateHabit(payload: UpdateHabitRequest) {
  return apiRequest<Habit>("/api/habits/update", {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export function deleteHabit(habitId: number) {
  return apiRequest<void>(`/api/habits/delete?habit_id=${habitId}`, {
    method: "DELETE",
  });
}

File: backend/src/routers/habits.py
from typing import List
from urllib import request

from fastapi import APIRouter, Request, Depends, HTTPException, Response
from sqlalchemy.orm import Session

from ..crud import create_habit, get_habit, get_habit_list, update_habit, delete_habit
from ..database import get_db
from ..schemas import HabitCreate, HabitRead, HabitUpdate

habits_router = APIRouter(
    prefix="/api/habits",
    tags=["habits"]
)


@habits_router.post("/create", response_model=HabitRead)
def create(body: HabitCreate,request: Request, db: Session = Depends(get_db)):
    return create_habit(db=db, payload=body, request=request)


@habits_router.get("/", response_model=List[HabitRead])
def read_habit(db:Session = Depends(get_db)):
    return get_habit_list(db=db)


@habits_router.patch("/update", response_model=HabitRead)
def update(request: HabitUpdate,db: Session = Depends(get_db)):
    return update_habit(db=db, payload=request)


@habits_router.delete("/delete", response_model=HabitRead)
def delete(request: Request,habit_id: int, db: Session = Depends(get_db)):
    return delete_habit(db=db,habit_id=habit_id)